src/hotspot/share/prims/agentList.cpp

```diff
@@ -0,0 +1,263 @@
+}
+
+AgentList::Iterator AgentList::agents() {
+  return Iterator(&_list, Iterator::NOT_XRUN);
+}
+
+AgentList::Iterator AgentList::java_agents() {
+  return Iterator(&_list, Iterator::JAVA);
+}
+
+AgentList::Iterator AgentList::native_agents() {
+  return Iterator(&_list, Iterator::NATIVE);
+}
+
+AgentList::Iterator AgentList::xrun_agents() {
+  return Iterator(&_list, Iterator::XRUN);
+}
+
+AgentList::Iterator AgentList::all() {
+  return Iterator(&_list, Iterator::ALL);
+}
+
+void AgentList::add(Agent* agent) {
+  assert(agent != nullptr, "invariant");
+  Agent* next;
+  do {
+    next = head(&_list);
+    agent->set_next(next);
+  } while (Atomic::cmpxchg(&_list, next, agent) != next);
+}
+
+void AgentList::add(const char* name, char* options, bool absolute_path) {
+  add(new Agent(name, options, absolute_path));
+}
+
+void AgentList::add_xrun(const char* name, char* options, bool absolute_path) {
+  Agent* agent = new Agent(name, options, absolute_path);
+  agent->set_xrun();
+  add(agent);
+}
+
+#ifdef ASSERT
+static void assert_initialized(AgentList::Iterator& it) {
+  while (it.has_next()) {
+    assert(it.next()->is_initialized(), "invariant");
+  }
+}
+#endif
+
+// In case an agent did not enable the VMInit callback, or if it is an -Xrun agent,
+// it gets an initializiation timestamp here.
+void AgentList::initialize() {
+  Iterator it = all();
+  while (it.has_next()) {
+    Agent* agent = it.next();
+    if (!agent->is_initialized()) {
+      agent->initialization_begin();
+    }
+  }
+  DEBUG_ONLY(Iterator assert_it = all(); assert_initialized(assert_it);)
+}
+
+void AgentList::convert_xrun_agents() {
+  Iterator it = xrun_agents();
+  while (it.has_next()) {
+    it.next()->convert_xrun_agent();
+  }
+}
+
+class JvmtiPhaseTransition : public StackObj {
+ public:
+  JvmtiPhaseTransition() {
+    assert(JvmtiEnvBase::get_phase() == JVMTI_PHASE_PRIMORDIAL, "invalid init sequence");
+    JvmtiExport::enter_onload_phase();
+  }
+  ~JvmtiPhaseTransition() {
+    assert(JvmtiEnvBase::get_phase() == JVMTI_PHASE_ONLOAD, "invariant");
+    JvmtiExport::enter_primordial_phase();
+  }
+};
+
+static void load_agents(AgentList::Iterator& it) {
+  while (it.has_next()) {
+    it.next()->load();
+  }
+}
+
+// Invokes Agent_OnLoad for -agentlib:.. -agentpath:  and converted -Xrun agents.
+// Called very early -- before JavaThreads exist
+void AgentList::load_agents() {
+  // Convert -Xrun to -agentlib: if there is no JVM_OnLoad
+  convert_xrun_agents();
+  JvmtiPhaseTransition transition;
+  Iterator it = agents();
+  ::load_agents(it);
+}
+
+// Launch -Xrun agents
+void AgentList::load_xrun_agents() {
+  assert(JvmtiEnvBase::get_phase() == JVMTI_PHASE_PRIMORDIAL, "invalid init sequence");
+  Iterator it = xrun_agents();
+  ::load_agents(it);
+}
+
+// Invokes Agent_OnAttach for agents loaded dynamically during runtime.
+jint AgentList::load_agent(const char* agent_name, const char* absParam,
```

Comment:
I feel that it is better to keep the original function name "load_agent_library". As you listed there two kinds of agents: Java and Native. The function name give a hint it is native agent. Also, it is better to avoid changes that aren't really necessary.

Reply:
I changed the names because I found it very hard to understand what the old names represented: "AgentLibrary" vs "Library"? "add_init_agent" vs "add_instrumentation_agent", or even "add_loaded_agent"? Also a bit confusing that "load_agent_library" would also include statically linked agents - no library is loaded there.

Reply:
Refactoring is usually not easy to review.
With a renaming it becomes harder, so it is better to be conservative.

There are other side effects to consider:

  - back porting also becomes harder
  - developers have to learn new names instead of already known
  
The good side is that your refactoring consolidates this code in a well known locations. :-)

Reply:
Of course, I would not have changed this unless I believe it improves things. The abstraction is now better from the perspective of the rest of the VM. There are now only JVMTI agents, and they are kept in a list. Arguments.cpp adds agents to the list. The same thing for the diagnosticCommand.cpp for dynamically loaded agents. Threads.cpp loads the JVMTI agents, java.cpp unloads agents. All other sites take out an iterator of the subtypes they want to iterate.

There is no longer any separation between "agent" and "library"; the subtypes of the agents are now abstracted away.